torch/utils/data/datapipes/map/grouping.py

```diff
@@ -2,12 +2,76 @@
 
 from functools import reduce
 from torch.utils.data import MapDataPipe, functional_datapipe, DataChunk
-from typing import List, TypeVar, Union, Optional
+from typing import List, Optional, Sized, TypeVar, Union
 
 
 T = TypeVar('T')
 
 
+@functional_datapipe('batch')
+class BatcherMapDataPipe(MapDataPipe[DataChunk]):
+    r""" :class:`BatcherMapDataPipe`.
+
+    Map DataPipe to create mini-batches of data. An outer dimension will be added as
+    `batch_size` if `drop_last` is set to `True`, or `length % batch_size` for the
+    last batch if `drop_last` is set to `False`.
+
+    Args:
+        datapipe: Iterable DataPipe being batched
+        batch_size: The size of each batch
+        drop_last: Option to drop the last batch if it's not full
+        unbatch_level: Specifies if it necessary to unbatch source data before
+            applying new batching rule
+    """
+    datapipe: MapDataPipe
+    batch_size: int
+    drop_last: bool
+    length: Optional[int]
+
+    def __init__(self,
+                 datapipe: MapDataPipe[T],
+                 batch_size: int,
+                 drop_last: bool = False,
+                 unbatch_level: int = 0,
+                 wrapper_class=DataChunk,
+                 ) -> None:
+        assert batch_size > 0, "Batch size is required to be larger than 0!"
+        super().__init__()
+        if unbatch_level == 0:
+            self.datapipe = datapipe
+        else:
+            self.datapipe = datapipe.unbatch(unbatch_level=unbatch_level)
+        self.unbatch_level = unbatch_level
+        self.batch_size = batch_size
+        self.drop_last = drop_last
+        self.length = None
+        self.wrapper_class = wrapper_class
+
+    def __getitem__(self, index) -> DataChunk:
+        batch: List = []
+        indices = range(index * self.batch_size, (index + 1) * self.batch_size)
+        try:
+            for i in indices:
+                batch.append(self.datapipe[i])
+            return self.wrapper_class(batch)
+        except IndexError:
+            if not self.drop_last and len(batch) > 0:
+                return self.wrapper_class(batch)
+            else:
+                raise IndexError(f"Index {index} is out of bound.")
+
+    def __len__(self) -> int:
+        if self.length is not None:
+            return self.length
+        if isinstance(self.datapipe, Sized) and self.unbatch_level == 0:
+            if self.drop_last:
+                self.length = len(self.datapipe) // self.batch_size
+            else:
+                self.length = (len(self.datapipe) + self.batch_size - 1) // self.batch_size
+            return self.length
+        raise TypeError("{} instance doesn't have valid length".format(type(self).__name__))
```

Comment:
IMO, we should assume `__len__` is always implemented for `MapDataPipe`

Reply:
I agree, though there are cases where it can be hard to compute (especially when one element in the input DataPipe can expands into many).

Reply:
The unbatch_level is super annoying here. Otherwise, we don't need to check if self.datapipe is Sized.

Reply:
I think you still need to check for it before you call `len(self.datapipe)` right?

In either case, maybe we should get rid of `unbatch_level`? I feel that users can call `.unbatch` prior to this if they wanted that operation. I decided to include it only because `IterDataPipe`'s version has that argument and I hesitate to make the two different.

Reply:
If we assume prior datapipe has `__len__`, then we can directly call `len(self.datapipe`.

> In either case, maybe we should get rid of unbatch_level? I feel that users can call .unbatch prior to this if they wanted that operation. I decided to include it only because IterDataPipe's version has that argument and I hesitate to make the two different.

Yeah. I feel like `unbatch_level` making me confused even for IterDataPipe. At least based on my experience with Vision and Text, no one really uses it.

Reply:
If that is the case, I think we should remove it for simplicity. I will make a PR and let the domains users have a quick look. I agree that it is unlikely that people are using it.

Reply:
Sound great. Thanks